File: src/python/xerxes/acp/runner.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Bridge the ACP server surface to Xerxes's real streaming agent loop.

:class:`AcpAgentRunner` bootstraps a :class:`RuntimeManager` (provider,
system prompt, tools, skills — exactly what the daemon uses) and turns an
ACP ``prompt`` into a real turn through :func:`xerxes.streaming.loop.run`.
Each :class:`AcpSession` keeps its own :class:`AgentState`, so multi-turn
conversations persist per session. Stream events are converted to ACP wire
events and pushed through the caller-supplied ``emit`` callback; tool
approvals route through the server's :class:`AcpPermissionBoard`."""

from __future__ import annotations

import logging
import threading
import uuid
from collections.abc import Callable
from typing import Any

from ..streaming.events import AgentState, PermissionRequest, TurnDone
from ..streaming.loop import run as run_loop
from .events import to_acp_event
from .permissions import AcpPermissionBoard, route_permission
from .session import AcpSession

logger = logging.getLogger(__name__)

EmitFn = Callable[[dict[str, Any]], None]


class AcpAgentRunner:
    """Drive real agent turns for the ACP server.

    Holds one bootstrapped :class:`RuntimeManager` for the process and a
    per-session :class:`AgentState`. ``permission_board`` is wired by the
    server so interactive approvals can block until the client responds."""

    def __init__(self, *, default_permission_mode: str = "auto") -> None:
        """Bootstrap the runtime once (provider/tools/skills) from the active profile."""
        from ..daemon.config import load_config
        from ..daemon.runtime import RuntimeManager

        self._runtime = RuntimeManager(config=load_config())
        self._runtime.reload()
        self._states: dict[str, AgentState] = {}
        self._default_permission_mode = default_permission_mode
        self.permission_board: AcpPermissionBoard | None = None
        # Interactive ``AskUserQuestion`` routing: a question board + per-thread
        # turn context so the (global, blocking) tool callback can find the
        # current session/emit and surface the question to the ACP client.
        self._questions: dict[str, dict[str, Any]] = {}
        self._q_lock = threading.Lock()
        self._ctx = threading.local()

    # ----- list providers (wired into AcpServer) ---------------------------

    def list_tools(self) -> list[dict[str, Any]]:
        """Return the live OpenAI-shaped tool schemas."""
        return list(self._runtime.tool_schemas)

    def list_models(self) -> list[dict[str, Any]]:
        """Return available model ids (provider catalogue, else the active model)."""
        from ..bridge import profiles

        cfg = self._runtime.runtime_config
        base_url = str(cfg.get("base_url", ""))
        models: list[str] = []
        if base_url:
            try:
                models = profiles.fetch_models(base_url, str(cfg.get("api_key", "")))
            except Exception as exc:
                logger.debug("fetch_models failed: %s", exc)
        if not models and self._runtime.model:
            models = [self._runtime.model]
        return [{"id": m, "name": m} for m in models]

    @property
    def active_model(self) -> str:
        """The runtime's active model id."""
        return self._runtime.model

    # ----- prompt execution -------------------------------------------------

    def run_prompt(self, *, session: AcpSession, text: str, emit: EmitFn | None = None, **_: Any) -> dict[str, Any]:
        """Run one turn for ``session`` and stream ACP events via ``emit``.

        Returns a summary dict (token counts / stop reason) once the turn
        ends. ``emit`` receives already-wire-shaped ACP event dicts; when
        omitted (pure-logic use) events are simply not streamed."""
        emit = emit or (lambda _event: None)
        state = self._states.setdefault(session.session_id, AgentState(messages=[]))

        config = dict(self._runtime.runtime_config)
        config["permission_mode"] = session.metadata.get("permission_mode", self._default_permission_mode)
        if session.model_override:
            config["model"] = session.model_override

        # Expose this turn's session/emit to the (global, blocking) AskUserQuestion
        # callback running on this same thread during tool execution.
        self._ctx.session = session
        self._ctx.emit = emit
        summary: dict[str, Any] = {"ok": True, "cancelled": False}
        try:
            for event in run_loop(
                user_message=text,
                state=state,
                config=config,
                system_prompt=self._runtime.system_prompt,
                tool_executor=self._runtime.tool_executor,
                tool_schemas=self._runtime.tool_schemas,
                cancel_check=lambda: session.cancelled,
            ):
                if isinstance(event, PermissionRequest):
                    event.granted = self._resolve_permission(session, event, emit)
                    # Still surface the request to the client for visibility.
                    continue
                emit(to_acp_event(event).to_wire())
                if isinstance(event, TurnDone):
                    summary.update(
                        {
                            "input_tokens": event.input_tokens,
                            "output_tokens": event.output_tokens,
                            "tool_calls_count": event.tool_calls_count,
                            "model": event.model,
                        }
                    )
        except Exception as exc:
            logger.warning("ACP prompt failed for session %s: %s", session.session_id, exc)
            return {"ok": False, "error": str(exc)}
        finally:
            self._ctx.session = None
            self._ctx.emit = None
        summary["cancelled"] = session.cancelled
        return summary

    def _resolve_permission(self, session: AcpSession, event: PermissionRequest, emit: EmitFn) -> bool:
        """Surface a permission request to the client and block until decided.

        Auto-denies on cancellation. Without a wired board (pure-logic mode)
        the request is auto-denied so the turn can't hang."""
        board = self.permission_board
        req = route_permission(
            session_id=session.session_id,
            tool_name=event.tool_name,
            description=event.description,
            inputs=event.inputs,
        )
        emit(
            {
                "kind": "permission_request",
                "permission_id": req.id,
                "session_id": session.session_id,
                "tool_name": event.tool_name,
                "description": event.description,
                "inputs": event.inputs,
            }
        )
        if board is None:
            return False
        waiter = board.submit(req)
        while not waiter.wait(timeout=0.2):
            if session.cancelled:
                board.drop(req.id)
                return False
        allowed = req.allowed
        board.drop(req.id)
        return bool(allowed and not session.cancelled)

    # ----- interactive AskUserQuestion (routed to the ACP client) ----------

    def ask_user_question(self, question: str) -> str:
        """Blocking callback for ``AskUserQuestionTool`` — ask the ACP client.

        Wired via ``set_ask_user_question_callback``. Runs on the prompt worker
        thread during tool execution, so it reads this turn's session/emit from
        the thread-local context, surfaces an ``input_request`` to the client,
        and blocks until ``respond_question`` resolves it (or the session is
        cancelled). Raises if no turn is active."""
        session = getattr(self._ctx, "session", None)
        emit = getattr(self._ctx, "emit", None)
        if session is None or emit is None:
            raise RuntimeError("AskUserQuestion called outside an active ACP turn")

        qid = uuid.uuid4().hex
        event = threading.Event()
        with self._q_lock:
            self._questions[qid] = {"event": event, "answer": "", "question": question, "session_id": session.session_id}
        emit(
            {
                "kind": "input_request",
                "input_id": qid,
                "session_id": session.session_id,
                "question": question,
            }
        )
        while not event.wait(timeout=0.2):
            if session.cancelled:
                with self._q_lock:
                    self._questions.pop(qid, None)
                return ""
        with self._q_lock:
            entry = self._questions.pop(qid, None)
        return str(entry["answer"]) if entry else ""

    def respond_question(self, input_id: str, answer: str) -> dict[str, Any]:
        """Resolve a pending ``input_request`` with the client's ``answer``."""
        with self._q_lock:
            entry = self._questions.get(input_id)
            if entry is None or entry["event"].is_set():
                return {"ok": False}
            entry["answer"] = answer
            entry["event"].set()
        return {"ok": True}

    def pending_questions(self) -> list[dict[str, Any]]:
        """Snapshot of input requests awaiting a client answer."""
        with self._q_lock:
            return [
                {"input_id": qid, "session_id": e["session_id"], "question": e["question"]}
                for qid, e in self._questions.items()
                if not e["event"].is_set()
            ]

    def reset_session(self, session_id: str) -> None:
        """Drop the cached conversation state for ``session_id``."""
        self._states.pop(session_id, None)


__all__ = ["AcpAgentRunner"]

File: xerxes/src/acp/runner.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import { errorMessage, type ToolExecutor } from '../executors/toolRegistry.js'
import { mergePersistedSubagentSnapshots } from '../agents/subagentPersistence.js'
import type { SubagentTurnCoordinator } from '../daemon/subagentCoordinator.js'
import { formatSubagentResults } from '../daemon/turnRunner.js'
import type { LlmClient } from '../llms/client.js'
import { createAgentState, type AgentState, type PermissionRequest, type StreamEvent } from '../streaming/events.js'
import { runTurn } from '../streaming/loop.js'
import type { PermissionBroker, PermissionDecision, PermissionMode, ToolPolicy } from '../streaming/permissions.js'
import type { ToolDefinition } from '../types/toolCalls.js'
import { routePermission, type AcpPermissionBoard } from './permissions.js'
import type { AcpSession } from './session.js'
import type { AcpEventEmitter, AcpModelInfo, AcpPromptRequest, AcpWireEvent } from './types.js'
import { toAcpEvent } from './events.js'

export interface AcpAgentRunnerOptions {
  readonly agentId?: string
  readonly defaultPermissionMode?: PermissionMode
  readonly llm: LlmClient
  readonly maxTokens?: number
  readonly model: string
  readonly modelListProvider?: () => readonly AcpModelInfo[]
  readonly policy?: ToolPolicy
  readonly systemPrompt?: string
  /** Joins detached child work back into the ACP prompt that created it. */
  readonly subagentCoordinator?: SubagentTurnCoordinator
  readonly temperature?: number
  readonly toolExecutor?: ToolExecutor
  readonly tools?: readonly ToolDefinition[]
  readonly topP?: number
}

interface ActivePrompt {
  readonly controller: AbortController
  readonly emit: AcpEventEmitter
  readonly session: AcpSession
}

interface PendingQuestion {
  readonly abortHandler: () => void
  readonly question: string
  readonly resolve: (answer: string) => void
  readonly sessionId: string
  readonly signal: AbortSignal
}

/**
 * Drive the portable TypeScript agent loop for ACP sessions.
 *
 * State is retained independently for each ACP session. The runner itself is
 * provider-neutral: callers inject an `LlmClient`, tool executor, schemas, and
 * policy just as they do for the daemon's `AgentTurnRunner`.
 */
export class AcpAgentRunner {
  private readonly activePrompts = new Map<string, ActivePrompt>()
  private readonly options: AcpAgentRunnerOptions
  private permissionBoard: AcpPermissionBoard | undefined
  private readonly questions = new Map<string, PendingQuestion>()
  private readonly states = new Map<string, AgentState>()

  constructor(options: AcpAgentRunnerOptions) {
    this.options = options
  }

  cancel(sessionId: string): boolean {
    const active = this.activePrompts.get(sessionId)
    if (!active) {
      return false
    }
    active.session.cancelled = true
    active.controller.abort(new Error('ACP prompt cancelled'))
    this.resolveQuestionsForSession(sessionId, '')
    return true
  }

  listModels(): readonly AcpModelInfo[] {
    return [...(this.options.modelListProvider?.() ?? [{ id: this.options.model, name: this.options.model }])]
  }

  listTools(): readonly ToolDefinition[] {
    return [...(this.options.tools ?? [])]
  }

  pendingQuestions(): readonly Record<string, unknown>[] {
    return [...this.questions.entries()].map(([inputId, question]) => ({
      input_id: inputId,
      session_id: question.sessionId,
      question: question.question,
    }))
  }

  resetSession(sessionId: string): void {
    this.states.delete(sessionId)
    this.resolveQuestionsForSession(sessionId, '')
  }

  respondQuestion(inputId: string, answer: string): Record<string, boolean> {
    const question = this.questions.get(inputId)
    if (!question) {
      return { ok: false }
    }
    this.questions.delete(inputId)
    question.signal.removeEventListener('abort', question.abortHandler)
    question.resolve(answer)
    return { ok: true }
  }

  setPermissionBoard(board: AcpPermissionBoard): void {
    this.permissionBoard = board
  }

  stateFor(sessionId: string): AgentState | undefined {
    return this.states.get(sessionId)
  }

  /** Run one streamed ACP prompt and return the final token/model summary. */
  async runPrompt(request: AcpPromptRequest): Promise<Record<string, unknown>> {
    const { session } = request
    if (this.activePrompts.has(session.sessionId)) {
      return { ok: false, error: `prompt already active for session: ${session.sessionId}` }
    }

    const controller = new AbortController()
    if (session.cancelled) {
      controller.abort(new Error('ACP session is cancelled'))
    }
    const emit = request.emit ?? (() => undefined)
    const active: ActivePrompt = { session, controller, emit }
    this.activePrompts.set(session.sessionId, active)
    const state = this.states.get(session.sessionId) ?? createAgentState()
    this.states.set(session.sessionId, state)
    const model = session.modelOverride ?? this.options.model
    const permissionMode = permissionModeFor(session, this.options.defaultPermissionMode ?? 'accept-all')
    state.metadata.permission_mode = permissionMode
    const summary: Record<string, unknown> = { ok: true, cancelled: false }
    const subagentCohort = this.options.subagentCoordinator?.begin(session.sessionId)

    try {
      const permissionBroker: PermissionBroker = {
        request: (permission, signal) => this.resolvePermission(session, permission, emit, signal),
      }
      for await (const event of runTurn({
        model,
        state,
        userMessage: request.text,
        sessionId: session.sessionId,
        ...(this.options.agentId ? { agentId: this.options.agentId } : {}),
        ...(this.options.maxTokens !== undefined ? { maxTokens: this.options.maxTokens } : {}),
        ...(this.options.systemPrompt ? { systemPrompt: this.options.systemPrompt } : {}),
        ...(this.options.temperature !== undefined ? { temperature: this.options.temperature } : {}),
        ...(this.options.tools ? { tools: this.options.tools } : {}),
        ...(this.options.topP !== undefined ? { topP: this.options.topP } : {}),
        permissionMode,
      }, {
        ...(subagentCohort ? {
          awaitAgentEvents: async signal => {
            const snapshots = await subagentCohort.waitForResults(signal)
            mergePersistedSubagentSnapshots(state.metadata, snapshots)
            return formatSubagentResults(snapshots)
          },
        } : {}),
        llm: this.options.llm,
        permissionBroker,
        ...(this.options.policy ? { policy: this.options.policy } : {}),
        ...(this.options.toolExecutor ? { toolExecutor: this.options.toolExecutor } : {}),
      }, controller.signal)) {
        if (event.type === 'permission_request') {
          continue
        }
        await emit(toAcpEvent(event).toWire())
        updateSummary(summary, event)
      }
    } catch (error) {
      return {
        ok: false,
        error: errorMessage(error),
        cancelled: session.cancelled || controller.signal.aborted,
      }
    } finally {
      subagentCohort?.close()
      this.activePrompts.delete(session.sessionId)
      this.resolveQuestionsForSession(session.sessionId, '')
    }

    summary.cancelled = session.cancelled || controller.signal.aborted
    return summary
  }

  /**
   * Surface a tool's question to the ACP client and await its answer.
   *
   * Tool adapters that need interactive input can call this method with their
   * current ACP session id; it deliberately has no process-global callback.
   */
  async askUserQuestion(sessionId: string, question: string): Promise<string> {
    const active = this.activePrompts.get(sessionId)
    if (!active) {
      throw new Error('AskUserQuestion called outside an active ACP turn')
    }
    if (active.controller.signal.aborted || active.session.cancelled) {
      return ''
    }

    const inputId = crypto.randomUUID().replaceAll('-', '')
    return new Promise<string>((resolve, reject) => {
      const abortHandler = () => {
        this.questions.delete(inputId)
        resolve('')
      }
      const pending: PendingQuestion = {
        sessionId,
        question,
        resolve,
        abortHandler,
        signal: active.controller.signal,
      }
      this.questions.set(inputId, pending)
      active.controller.signal.addEventListener('abort', abortHandler, { once: true })
      if (active.controller.signal.aborted) {
        abortHandler()
        return
      }
      void Promise.resolve(active.emit({
        kind: 'input_request',
        input_id: inputId,
        session_id: sessionId,
        question,
      })).catch(error => {
        this.questions.delete(inputId)
        active.controller.signal.removeEventListener('abort', abortHandler)
        reject(error)
      })
    })
  }

  private async resolvePermission(
    session: AcpSession,
    permission: PermissionRequest,
    emit: AcpEventEmitter,
    signal?: AbortSignal,
  ): Promise<PermissionDecision> {
    const request = routePermission({
      sessionId: session.sessionId,
      toolName: permission.toolCall.function.name,
      description: permission.description,
      inputs: permission.inputs,
    })
    const board = this.permissionBoard
    if (!board) {
      await emit({
        kind: 'permission_request',
        permission_id: request.id,
        session_id: session.sessionId,
        tool_name: request.toolName,
        description: request.description,
        inputs: request.inputs,
      })
      return 'reject'
    }
    // Register before notifying the client so a fast `permission/respond`
    // cannot race ahead of the pending-request board.
    board.submit(request)
    try {
      await emit({
        kind: 'permission_request',
        permission_id: request.id,
        session_id: session.sessionId,
        tool_name: request.toolName,
        description: request.description,
        inputs: request.inputs,
      })
      const allowed = await board.awaitDecision(request, signal)
      return allowed && !session.cancelled && !signal?.aborted ? 'approve' : 'reject'
    } finally {
      board.drop(request.id)
    }
  }

  private resolveQuestionsForSession(sessionId: string, answer: string): void {
    for (const [inputId, question] of this.questions) {
      if (question.sessionId !== sessionId) {
        continue
      }
      this.questions.delete(inputId)
      question.signal.removeEventListener('abort', question.abortHandler)
      question.resolve(answer)
    }
  }
}

function permissionModeFor(session: AcpSession, fallback: PermissionMode): PermissionMode {
  const raw = session.metadata.permission_mode ?? session.metadata.permissionMode
  return raw === 'accept-all' || raw === 'auto' || raw === 'manual' || raw === 'plan' ? raw : fallback
}

function updateSummary(summary: Record<string, unknown>, event: StreamEvent): void {
  if (event.type !== 'turn_done') {
    return
  }
  summary.input_tokens = event.usage.inputTokens
  summary.output_tokens = event.usage.outputTokens
  summary.tool_calls_count = event.toolCallsCount
  summary.model = event.model
}
